package/src/utilities/add-devtoolbar-framework-app.ts

```diff
@@ -0,0 +1,131 @@
+};
+
+/**
+ * Add a Dev Toolbar Plugin that uses a Framework component.
+ *
+ * @param {object} params
+ * @param {string} params.name - The name of the toolbar plugin
+ * @param {string} params.icon - This should be an inline SVG
+ * @param {URL} params.framework
+ * @param {URL} params.src - Path to your component
+ * @param {URL} params.style - A stylesheet to pass to your plugin
+ * @param {URL} params.callback - A callback function containing the canvas and window your plugin is loaded on
+ * @param {import("astro").HookParameters<"astro:config:setup">["updateConfig"]} params.updateConfig
+ * @param {import("astro").HookParameters<"astro:config:setup">["addDevToolbarApp"]} params.addDevToolbarApp
+ * @param {import("astro").HookParameters<"astro:config:setup">["injectScript"]} params.injectScript
+ *
+ * @example
+ * ```ts
+ * addDevToolbarFrameworkApp({
+ *      framework: "vue",
+ *      name: "Test Vue Plugin",
+ *      id: "test-vue-plugin",
+ *      icon: `<svg version="1.1" viewBox="0 0 261.76 226.69" xmlns="http://www.w3.org/2000/svg"><g transform="matrix(1.3333 0 0 -1.3333 -76.311 313.34)"><g transform="translate(178.06 235.01)"><path d="m0 0-22.669-39.264-22.669 39.264h-75.491l98.16-170.02 98.16 170.02z" fill="#41b883"/></g><g transform="translate(178.06 235.01)"><path d="m0 0-22.669-39.264-22.669 39.264h-36.227l58.896-102.01 58.896 102.01z" fill="#34495e"/></g></g></svg>`,
+ *      src: resolve("./Test.vue"),
+ *      style: `
+ *          button {
+ *              background-color: rebeccapurple;
+ *          }
+ *      `,
+ * });
+ * ```
+ *
+ * @see https://astro-integration-kit.netlify.app/utilities/add-devtoolbar-plugin/
+ */
+export const addDevToolbarFrameworkApp = ({
+	id,
+	name,
+	icon,
+	framework,
+	src,
+	style,
+	addDevToolbarApp,
+	updateConfig,
+	injectScript,
+}: addDevToolbarFrameworkAppParams) => {
+	const virtualModuleName = `virtual:astro-devtoolbar-app-${id}`;
+
+	const missingImportsForFramework = missingImports[framework];
+	if (missingImportsForFramework.length > 0) {
+		throw new AstroError(
+			`Missing dependencies for ${framework} framework: ${missingImportsForFramework.join(
+				", ",
+			)}`,
+		);
+	}
+
+	const { resolve } = createResolver(import.meta.url);
+
+	let content = readFileSync(
+		resolve(`./addDevToolbarFrameworkAppStubs/${framework}.ts`),
+		"utf8",
+	);
+
+	const escapedIcon = icon
+		.replace("`", '${"`"}')
+
+	content = content
+		.replace("@@COMPONENT_SRC@@", src)
+		.replace("@@ID@@", id)
+		.replace("@@NAME@@", name)
+		.replace("@@ICON@@", escapedIcon)
+		.replace("@@STYLE@@", style ?? "");
+
+	addVirtualImport({
+		name: virtualModuleName,
+		content,
+		updateConfig,
+	});
+
+	switch (framework) {
+		case "react":
+			import("@vitejs/plugin-react").then((react) => {
+				const FAST_REFRESH_PREAMBLE = react.default.preambleCode;
+				const preamble = FAST_REFRESH_PREAMBLE.replace("__BASE__", "/");
+				injectScript("page", preamble);
+			});
+
+			break;
+	}
```

Comment:
hasn't been so far, because all this will run in the setup it should be fine. I only did this because I async/sync stuff, so I didn't want to have an await in here... But actually Fry was saying all imports are secretly awaited or something like that 🤔 I don't know, I don't think it's ever going to be a problem this way, but happy to change as long as it doesn't mess it up

Reply:
all fine let's keep the import().then! However my suggestion still stands

Reply:
> But actually Fry was saying all imports are secretly awaited

All import _statements_ are implicitly awaited.
```ts
// This hides an await point:
import foo from 'foo';

// This requires an explicit await point:
const foo = await import('foo');
```

Reply:
But this won't be a problem because that import is immediately available. The dependency check at the top already awaits this import, so it returns an already-resolved Promise. This callback is guaranteed to run in the current phase of the event loop.

Reply:
ah that's why, gotcha! thanks wizard